### PR TITLE
[ hnrm2 ] Use precision-enhanced hnrm2

### PR DESCRIPTION
**Pull request by skykongkong8**
Along with #2555 I inspected through every custom-made half-precision calculation function, and `hnrm2` is the only one left that needs f16-f32 precision supplement.

**Proposed Changes**
- Previous hnrm2 was using full-fp16.
- Since this is also one of dimension-shrinking computation, should use inter-fp32 values to enhance precision.
- This has not been detected due to small dimension Tensor usage in unittest. Add higher dimension test case accordingly.
- Note that this function is responsible for `Tensor::l2norm()`, frequently used for mse loss computation.

### Differnece w.r.t. fp32-cblas
| dim | f16 (prev) | f16f32 (now) |
| --- | --- | --- |
| 768 | 282.233 | 0.232574 |

| dim | f16 | f16f32 | fp32-cblas | 
| --- | --- | --- | --- |
| 768 | 508 ns | 579 ns | 4242 ns |

Better accuracy with almost no latency deterioration

**Self evaluation:**
1. Build test:     [X]Passed [ ]Failed [ ]Skipped
2. Run test:     [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

**Review by jijoongmoon — APPROVED** [review]
Great ! LGTM

File: nntrainer/tensor/blas_interface.cpp [hunk 1]
```diff
@@ -281,24 +281,28 @@ void sscal(const unsigned int N, const float alpha, _FP16 *X, const int incX) {
...
   _FP16 tmp;
 #if (defined USE__FP16 && USE_NEON)
   if (incX == 1) {
     sum = nntrainer::neon::hnrm2(N, X);
   } else {
+    float sum32 = 0;
     for (unsigned int i = 0; i < N; i++) {
       tmp = X[i * incx];
-      sum += tmp * tmp;
+      sum32 += tmp * tmp;
```
Comment (djeong20) [inline nntrainer/tensor/blas_interface.cpp:293]: just wondering. would having `tmp` a float instead of fp16 make a difference?

File: test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp [hunk 2]
```diff
@@ -150,11 +150,54 @@ TEST(nntrainer_Tensor, l2norm) {
...
+
+  __fp16 result_neon;
+  float result_fp32;
+
+  result_neon = input.l2norm();
+  result_fp32 = input_fp32.l2norm();
+
+  float ErrorNeon = abs(result_neon - result_fp32);
+
+  const float epsilon = 1e-3 * width;
```
Comment (djeong20) [inline test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp:197]: what's the reasoning for multiplying width? is it to sum errors of num of width?

File: test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp [hunk 3]
```diff
@@ -150,11 +150,54 @@ TEST(nntrainer_Tensor, l2norm) {
...
+                              MOD) *
+                               alpha);
+
+  __fp16 result_neon;
+  float result_fp32;
+
+  result_neon = input.l2norm();
+  result_fp32 = input_fp32.l2norm();
+
+  float ErrorNeon = abs(result_neon - result_fp32);
```
Comment (djeong20) [inline test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp:195]: maybe use camelCase or snake_case?
```suggestion
  float error_neon = abs(result_neon - result_fp32);
```

**Review by djeong20 — APPROVED** [review]
Nice work.

File: nntrainer/tensor/blas_interface.cpp (hunk 1 above)
Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp:293]: Definitely yes, for several reasons.

1. You know, fp16 arithmetic can only hold up to 65504, so using fp16 for summing variable(?) is not really a good idea.
It is highly likely to make overflow if N is sufficiently big number.
2. For the same reason, it might cause a underflow as well. For example, it might just skip small decimal values into consideration. As far as I am concerned, even with `__fp16` * `__fp16` multiplication, compiler internally changes them into `float`. Thus, storing that value in `__fp16` variable is also not a clever way I guess. At glance, the effect might seem trivial, but eventually it was causing quite big error when compared to output of fp32-cblas computation.

File: test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp (hunk 3 above)
Comment (skykongkong8) [reply test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp:195]: Maybe for the next PR? That is not really a point of this commit, but I will definitely take that in my mind.

File: test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp (hunk 2 above)
Comment (skykongkong8) [reply test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp:197]: There is unavoidable round-off error that linearly increase w.r.t. tensor size in half-precision computation. Meanwhile, I am comparing half-precision result with fp32 result. If they were using the same precision, I could use mathematically-defined criteria, but it is not. So I have been using this kind of epsilon to compensate such error.
Strictly speaking, this function makes (M x N) Matrix to one single value, so something like:
```C++
  const float epsilon = 1e-3 * height * width;
  // or const float epsilon = sqrt(1e-3 * height * width); ???
``` 
would be better in terms of consistency, but I just left it this way this way for stricter condition.

File: nntrainer/tensor/blas_interface.cpp (hunk 1 above)
Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp:293]: Ah, I misunderstood. Thought you were asking about `sum32`.
But like I mentioned above, the compiler will internally compute them with `float` anyway, so think it won't make huge difference I guess.
Additionally, personally think this is better way i terms of consistency as well. SIMD functions for this computation is implemented is like : "multiply with f16 - and accumulate with f32"

File: test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp (hunk 2 above)
Comment (djeong20) [reply test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp:197]: makes sense. if so, I think comparing fp16 with fp32 could be pointless since it would never be close (especially when the dimension increases). how about we compare the fp16 tensor l2 norm value with other frameworks such as PyTorch or Tensorflow?

Comment (skykongkong8) [reply test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp:197]: Nice approach...
Then, is there a efficient way to compare Pytorch output and ourcase in unittest codes?
I don't really prefer hard-coded unittest cases like we do now...

Comment (djeong20) [reply test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp:197]: I tested the added test case, `l2norm_big768`, with PyTorch and Tensorflow. The L2 norm value for the float32 tensor matches with both frameworks. However, there's a mismatch for float16.

#### NNTrainer
Before: 45.25
After: 410.25

#### PyTorch
156.7500

#### Tensorflow
inf

This fix gets the closest value to the float32 tensor, which is 409.999, so there's no need to compare with other frameworks. I think it is good to go :)

Comment (skykongkong8) [reply test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp:197]: Based on the output from PyTorch / Tensorflow, I think they are just computing the value with full-fp16 precision.
Mathematically it is not really a wrong value, but it is not actually desired way in our framework I guess.
Anyway, thanks for pointing this issue out! 👍

**Review by baek2sm — APPROVED** [review]
LGTM
